pkg/provider/acme/local_store.go

```diff
@@ -31,10 +31,16 @@ func NewLocalStore(filename string) *LocalStore {
 
 func (s *LocalStore) save(resolverName string, storedData *StoredData) {
 	s.lock.Lock()
-	defer s.lock.Unlock()
-
 	s.storedData[resolverName] = storedData
-	s.saveDataChan <- s.storedData
+	s.lock.Unlock()
+
+	s.lock.RLock()
+	storedDataCopy := s.unSafeCopyOfStoredData(s.storedData)
+	s.lock.RUnlock()
+
+	// we cannot pass s.storedData directly, map is reference type and as result
+	// we can face with race condition, so we need to work with objects copy
+	s.saveDataChan <- storedDataCopy
```

Comment:
To follow up my first comment.

```suggestion

	// we cannot pass s.storedData directly, map is reference type and as result
	// we can face with race condition, so we need to work with objects copy
	s.saveDataChan <- s.copyOfStoredData()
```